cradle-core/src/main/java/com/exactpro/cradle/BookInfo.java

```diff
@@ -84,102 +148,247 @@ public Instant getCreated()
-		if (this == o) return true;
-		if (o == null || getClass() != o.getClass()) return false;
-		BookInfo bookInfo = (BookInfo) o;
+	private static Instant toInstant(long epochDay) {
+		return  Instant.ofEpochSecond(epochDay * SECONDS_IN_DAY);
+	}
+
+	@FunctionalInterface
+	public interface PageLoader {
+		@Nullable
+		PageInfo load(BookId bookId);
+	}
+
+	private interface IPageInterval {
+		PageInfo get(PageId pageId);
+
+		PageInfo find(Instant timestamp);
 
-		if (!Objects.equals(this.getId(), bookInfo.getId())) {
-			return false;
+		PageInfo next(Instant startTimestamp);
+
+		PageInfo previous(Instant startTimestamp);
+
+		void remove(PageId pageId);
+
+		void add(PageInfo page);
+	}
+
+	private static final class EmptyPageInterval implements IPageInterval {
+		@Override
+		public PageInfo get(PageId pageId) {
+			return null;
 		}
-		if (!Objects.equals(this.getCreated(), bookInfo.getCreated())) {
-			return false;
+
+		@Override
+		public PageInfo find(Instant timestamp) {
+			return null;
 		}
 
-		if (!Objects.equals(this.getDesc(), bookInfo.getDesc())) {
-			return false;
+		@Override
+		public PageInfo next(Instant startTimestamp) {
+			return null;
 		}
 
-		if (!Objects.equals(this.getFullName(), bookInfo.getFullName())) {
-			return false;
+		@Override
+		public PageInfo previous(Instant startTimestamp) {
+			return null;
 		}
 
-		/*
-			As getPages returns new unmodifiable list of ordered pages
-			this should work without any problems
-		 */
-		List<PageInfo> pages = new ArrayList<> (this.getPages());
-		List<PageInfo> otherPages = new ArrayList<> (bookInfo.getPages());
+		@Override
+		public void remove(PageId pageId) {
+			throw new UnsupportedOperationException();
+		}
+
+		@Override
+		public void add(PageInfo page) {
+			throw new UnsupportedOperationException();
+		}
+	}
+
+	private static class PageInterval implements IPageInterval {
+		private final Map<PageId, PageInfo> pageById = new HashMap<>();
+		private final TreeMap<Instant, PageInfo> pageByInstant = new TreeMap<>();
 
-		if (pages.size() != otherPages.size()) {
-			return false;
+        private PageInterval() { }
+
+		@Override
+		public PageInfo get(PageId pageId) {
+			return pageById.get(pageId);
 		}
 
-		for (int i = 0; i < pages.size(); i ++) {
-			if (!pages.get(i).equals(otherPages.get(i))) {
-				return false;
-			}
+		@Override
+		public PageInfo find(Instant timestamp)
+		{
+			Entry<Instant, PageInfo> result = pageByInstant.floorEntry(timestamp);
+			return result != null ? result.getValue() : null;
 		}
 
-		return true;
-	}
+		@Override
+		public PageInfo next(Instant startTimestamp) {
+			Entry<Instant, PageInfo> result = pageByInstant.ceilingEntry(startTimestamp.plus(1, ChronoUnit.NANOS));
+			return result != null ? result.getValue() : null;
+		}
+
+		@Override
+		public PageInfo previous(Instant startTimestamp) {
+			Entry<Instant, PageInfo> result = pageByInstant.floorEntry(startTimestamp.minus(1, ChronoUnit.NANOS));
```

Comment:
There are `higherKey` and `lowerKey` methods in `TreeMap`. They use strict comparison (`<` or `>`). Probably, we can use them here instead